### PR TITLE
refactor: Update nvidia-gpu-efa pattern

### PR DESCRIPTION
**Pull request by iankouls-aws**
# Description

### Motivation and Context

- Update the nvidia-gpu-efa pattern, enabling it to use an ODCR for capacity-challenged instance types.
- Update documentation
- Add EFA tests based on new public container

### How was this change tested?

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

### CODE REVIEW

File: examples/eks-efa/README.md [hunk 1]
```diff
@@ -0,0 +1,659 @@
+# EKS Blueprint Example with Elastic Fabric Adapter
```
Comment (bryantbiggs) [inline examples/eks-efa/README.md:1]: this directory no longer exists on `main` - it was converted into `patterns/nvidia-gpu-efa` so we can remove everything under this `examples/eks-efa`

File: patterns/nvidia-gpu-efa/README.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
-# EKS Cluster w/ NVIDIA GPUs and EFA for Machine Learning
+# EKS Cluster w/ NVIDIA GPUs and EFA for AI/ML Workloads
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/README.md:1]: lets revert - just machine learning

File: patterns/nvidia-gpu-efa/README.md [hunk 3]
```diff
@@ -17,46 +17,52 @@ The following components are demonstrated in this pattern:
 
 ## Code
 
-```terraform hl_lines="24-26 32-67"
-{% include  "../../patterns/nvidia-gpu-efa/eks.tf" %}
-```
+The code consists of the following files:
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/README.md:20]: please revert these changes - what is being provided is described above, this section is highlighting those key points for this pattern through the code (per the section title)

File: patterns/nvidia-gpu-efa/README.md [hunk 4]
```diff
@@ -17,46 +17,52 @@ The following components are demonstrated in this pattern:
...
+[main.tf](main.tf) - main project template, includes [`eks.tf`](eks.tf) and [`helm.tf`](helm.tf)
 
 ## Deploy
 
 See [here](https://aws-ia.github.io/terraform-aws-eks-blueprints/getting-started/#prerequisites) for the prerequisites and steps to deploy this pattern.
 
 ## Validate
 
-!!! note
+Note:
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/README.md:34]: please revert - this is for rendering correctly on doc site 
image

File: patterns/nvidia-gpu-efa/README.md [hunk 5]
```diff
@@ -82,84 +88,178 @@ See [here](https://aws-ia.github.io/terraform-aws-eks-blueprints/getting-started
...
 
-3. EFA test
+3. EFA info test
 
-    The results should shown that two EFA adapters are available (one for each worker pod)
+    This test prints a list of available EFA interfaces by using the `/opt/amazon/efa/bin/fi_info` utility.
+    Edit the script [generate-efa-info-test.sh](generate-efa-info-test.sh) and adjust the following environment variables if needed prior to running it:
+    
+    ```sh
+    export NUM_WORKERS=2
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/README.md:97]: the pattern is static so we don't use variables - we can educate and explain how users should adjust based on the instance type, and its subsequent resources, but the solution itself is static

File: patterns/nvidia-gpu-efa/README.md [hunk 6]
```diff
@@ -82,84 +88,178 @@ See [here](https://aws-ia.github.io/terraform-aws-eks-blueprints/getting-started
...
+    
+    `NUM_WORKERS` - number of nodes you want to run the test on
+    `GPU_PER_WORKER` - number of GPUs available on each node
+    `EFA_PER_WORKER` - number of EFA interfaces available on each node
+    
+    ```sh
+    ./generate-efa-info-test.sh
+    ```
+    
+    This script generates an MPIJob manifest file named `efa-info-test.yaml`.
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/README.md:110]: we should just provide the static yaml manifest, not a script to dynamically create one

File: patterns/nvidia-gpu-efa/README.md [hunk 7]
```diff
@@ -82,84 +88,178 @@ See [here](https://aws-ia.github.io/terraform-aws-eks-blueprints/getting-started
...
+    
+    ```sh
+    kubectl delete -f ./efa-info-test.yaml
+    ```
 
-    To run the EFA NCCL test please execute the following kubectl command:
+4. EFA NCCL test
 
+    The EFA NCCL test is used to measure network bandwidth by running the `/opt/nccl-tests/build/all_reduce_perf` utility. 
+    Please edit the script below and modify the environment variables at the beginning of the script as needed.
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/README.md:184]: likewise, make it static - here we are using a `p5.48xlarge` so those values are known

File: patterns/nvidia-gpu-efa/README.md [hunk 8]
```diff
@@ -82,84 +88,178 @@ See [here](https://aws-ia.github.io/terraform-aws-eks-blueprints/getting-started
...
+    ```sh
+    kubectl delete -f ./efa-nccl-test.yaml
     ```
 
 ## Destroy
 
-{%
-   include-markdown "../../docs/_partials/destroy.md"
-%}
+```sh
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/README.md:263]: revert this so the correct details are used for the doc site

File: patterns/nvidia-gpu-efa/eks.tf [hunk 9]
```diff
@@ -35,6 +35,15 @@ module "eks" {
       ami_type       = "AL2_x86_64_GPU"
       instance_types = ["p5.48xlarge"]
 
+
+      # Uncomment this block and specify capacity_reservation_id
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/eks.tf:39]: remove this - we have a dedicated example that shows the recommended usage https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/patterns/targeted-odcr

File: patterns/nvidia-gpu-efa/generate-efa-info-test.sh [hunk 10]
```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+
+export MPI_JOB_NAME=efa-info-test
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/generate-efa-info-test.sh:3]: these are all known values so lets just make it a static yaml file that users can apply

File: patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh [hunk 11]
```diff
@@ -0,0 +1,141 @@
+#!/bin/bash
+
+export MPI_JOB_NAME=efa-nccl-test
+export IMAGE_URI=public.ecr.aws/hpc-cloud/nccl-tests:latest
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh:4]: likewise for this manifest

File: patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+export INSTANCE_TYPE=p5.48xlarge
+export NUM_WORKERS=2
+export GPU_PER_WORKER=8
+export EFA_PER_WORKER=32
+export TOTAL_GPUS=$((${NUM_WORKERS}*${GPU_PER_WORKER}))
+
+export NCCL_NVLS_ENABLE=1
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh:11]: most of these can/should/must be removed https://github.com/aws/aws-ofi-nccl/blob/master/doc/efa-env-var.md

File: patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+            - PATH
+            - -x
+            - LD_LIBRARY_PATH
+            - -x
+            - XLA_FLAGS
+            - -x
+            - TF_XLA_FLAGS
+            - -x
+            - NCCL_DEBUG=INFO
+            #- -x
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh:78]: lets remove any of the commented out code

File: patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+                hugepages-2Mi: ${HUGEPAGES_2MI}
+                vpc.amazonaws.com/efa: ${EFA_PER_WORKER}
+                memory: ${MEMORY}
+              requests:
+                nvidia.com/gpu: ${GPU_PER_WORKER}
+                hugepages-2Mi: ${HUGEPAGES_2MI}
+                vpc.amazonaws.com/efa: ${EFA_PER_WORKER}
+                memory: ${MEMORY}
+          volumes:
+          - name: shmem
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh:138]: we are using shared memory or not? https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/1966/files#diff-ad70f23fba6a87f2255bd64844bd763630ca4f67b559b37e09a65ed24b0415a9R17

File: patterns/nvidia-gpu-efa/main.tf [hunk 15]
```diff
@@ -46,7 +46,7 @@ data "aws_availability_zones" "available" {}
 
 locals {
   name   = basename(path.cwd)
-  region = "us-west-2"
+  region = "us-east-2"
```
Comment (bryantbiggs) [inline patterns/nvidia-gpu-efa/main.tf:49]: revert this, all patterns default to `us-west-2`

**Review by bryantbiggs — CHANGES_REQUESTED** [review]

File: patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh (hunk 12 above)
Comment (mhuguesaws) [reply patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh:11]: Here is a ref for env parameter for NCCL, https://github.com/aws-samples/awsome-distributed-training/blob/main/micro-benchmarks/nccl-tests/kubernetes/nccl-tests.yaml#L40-L50

Comment (bryantbiggs) [reply patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh:11]: I'm more concerned that we are repeatedly setting a value that EFA states is harmful for these p4/p5 instances that we are using in the examples 

> RDMAV_FORK_SAFE=1
> 
>Do not use. This is a RDMA-core environment variable. Prefer FI_EFA_FORK_SAFE (if it still makes sense for your Linux kernel version). The two look the same, but actually behaves very differently, especially on newer kernels, where RDMAV_FORK_SAFE=1 can break things.

> RDMAV_*
> 
> Do not use.

File: examples/eks-efa/README.md (hunk 1 above)
Comment (iankouls-aws) [reply examples/eks-efa/README.md]: removed examples directory

File: patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh (hunk 12 above)
Comment (iankouls-aws) [reply patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh:11]: reconciled with manifest linked by mhuguesaws, removed commented out fields and RMDAV_*

File: patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh (hunk 14 above)
Comment (iankouls-aws) [reply patterns/nvidia-gpu-efa/generate-efa-nccl-test.sh:138]: yes using shmem by default, removed setting to make it clear

File: patterns/nvidia-gpu-efa/README.md (hunk 5 above)
Comment (iankouls-aws) [reply patterns/nvidia-gpu-efa/README.md:97]: I think the script makes sense here, but I agree that the user should not have to edit variables. 
I modified the instructions. Just by running the script, the default values will work. Users can easily adapt the example to other scenarios by setting different values later.

File: patterns/nvidia-gpu-efa/README.md (hunk 6 above)
Comment (iankouls-aws) [reply patterns/nvidia-gpu-efa/README.md]: The manifests are verbose and would make the readme harder to read, also copying and pasting yaml could introduce alignment errors. We could have the static yaml instead of script, but the difference is minimal. In my opinion, the script helps point out the different settings of the yaml where changes can occur and what their meaning is. The script works without change for two P5 instances, but can easily be adapted to other scenarios if needed, which is also valuable. Would like to keep this set up if acceptable.

File: patterns/nvidia-gpu-efa/eks.tf (hunk 9 above)
Comment (iankouls-aws) [reply patterns/nvidia-gpu-efa/eks.tf:39]: removed as requested

**Review by bryantbiggs — APPROVED** [review]
